Add UFACTORY 850 as one of the robots supported, introduce various HRI utilities for teleop and next episode hinting while reseting

vmayoral opened the pull request:
This PR comes as a result of the Oct 2024 Hackathon wherein I participated remotely. Results were successful within the weekend of hacking, so I thought I'd give back and contribute. Results first:

- [Setup with basic HRI for teleop](https://github.com/user-attachments/assets/5778900f-1e9d-4120-9f51-02d8acad3694)
- [`lerobot` teleop operational in UFACTORY 850](https://github.com/user-attachments/assets/aea3ac90-5b0b-4009-8dd6-047d6d7c800f)
- [Moving jellyfish grasped, `top`](https://github.com/user-attachments/assets/65699102-c78e-444c-8e85-6e738101e628)
- [Moving jellyfish grasped, `wrist`](https://github.com/user-attachments/assets/0a4a1a16-ad14-4058-ac77-1cbadb8a2753)

| resulting dataset |  resulting policy |
|-----|-----|
| https://huggingface.co/datasets/vmayoral/u850_test | https://huggingface.co/vmayoral/act_u850_test |

The contributions of this PR are summarized below:
- Add UFACTORY 850 as one of the supported robots, validing principles and introducing support for a commercial-grade *lowcost* cobot solution
- Enabled teleop while reseting episode 
- Various minor updates for usability including next episode dislosure while reseting
- Ensures datatypes are consistent across realsense camera instantiations
- Proposes a generic `xArmWrapper` of the UFACTORY Python SDK, usable with other arms from the same vendor
- extends `manipulator.py` with *minimal* changes (attempted respecting the already existing structure) to support the new arm tested
- Configs and related minor changes

Some notes and considerations:
- Current teleop mechanism in UFACTORY is jerky when compared to simpler implementations. There's plenty of margin for improvement here. Let me know if this of interest and I may find some room for contributing here as well.

vmayoral committed:
First steps towards integrating UFACTORY's 850 arm

- Added extensions to manipulator to support 850
- Create Motors' UFactory abstraction
- Proposed a new example demonstrating local setup

Signed-off-by: Víctor Mayoral Vilches <v.mayoralv@gmail.com>

vmayoral committed:
Minor updates in example following tutorial 7

Signed-off-by: Víctor Mayoral Vilches <v.mayoralv@gmail.com>

vmayoral committed:
Teleop with UFACTORY u850 adaptations

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

vmayoral committed:
Add UFACTORY's 850 config files

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

vmayoral committed:
Update UFACTORY's interface abstraction

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

vmayoral committed:
Add to reset_environment an intuition of next episode

This is useful while recording many episodes, since
you just loose track of them and the verbose outputs
from the teleop don't allow for clarity on this regard

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

vmayoral committed:
Final fixes before recording

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

vmayoral committed:
Fix some bugs in the port

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

vmayoral committed:
More bug fixes, adapt to new setup

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

vmayoral committed:
Re-organize examples

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

vmayoral committed:
Remove 12th example, since it's specific for 850

Also, it's a derivative of the previous ones

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

Cadene commented:
Wow really cool! We are in the middle of a refactor. We will come back to this PR soon. In the meantime, I am curious if other people working on UFACTORY 850 are able to use this arm with LeRobot :)

vmayoral commented:
Let me know @Cadene if you'd like other flavours of xArm robots tested. I'm rather certain it will, as the UFACTORY API is rather well structured and known reproducible. I've got an `xArm Lite 6` also lying around, happy to test it if that gets this PR moving forward.

Cadene commented:
@vmayoral Sounds good ;) Any chance you could run `pre-commit run --all-files`?
See: https://github.com/huggingface/lerobot/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr
Thanks 🙏

apockill commented on lerobot/common/robot_devices/robots/manipulator.py:
```diff
@@ -497,15 +515,21 @@ def teleop_step(
             # Slower fps expected due to reading from the follower.
             if self.config.max_relative_target is not None:
-                present_pos = self.follower_arms[name].read("Present_Position")
+                if self.robot_type == "u850":
```
This approach of adding if/elif throughout the codebase doesn't seem scalable.

What do folks think about creating an abstract base class that all robots can implement their wrapper API's within, so we don't have to edit so much code in a brittle way to add new robot support?

I can take this on- I'm currently working on adding support for the elephant robotics MyArmM&C series.

apockill replied:
Okay, sounds good! Maybe let's keep things as is, once xarm / myarm support is merged in I can jump in and write wrappers for all the currently supported robots.

The only hard thing will just be to have everyone test their bots out for regressions.

vmayoral committed:
Additional tests and example adjustments with u850

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

vmayoral committed:
Pre-commit alignment

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

vmayoral commented:
> @vmayoral Sounds good ;) Any chance you could run `pre-commit run --all-files`? See: https://github.com/huggingface/lerobot/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr Thanks 🙏

@Cadene all ready concerning these requests. Should be good to go as of right now. Just rebased on top of `main`.

vmayoral commented:
Ping @Cadene and team. Let me know if you folks require anything else.

vmayoral commented:
Friendly ping @Cadene. Let me know if you're missing anything.

vmayoral committed:
Add gello capabities to lerobot

Include an example called cooper.yaml which exemplifies the use
of Gello with a bimanual setup

Signed-off-by: vmayoral <v.mayoralv@gmail.com>

srisaripalli commented:
> 
> Some notes and considerations:
> 
>     * Current teleop mechanism in UFACTORY is jerky when compared to simpler implementations. There's plenty of margin for improvement here. Let me know if this of interest and I may find some room for contributing here as well.

@vmayoral  Thank you for this! We got the entire thing working on two xARM 850's. As you mentioned the teleop is quite jerky via control_robot.py or the 1_teleop_lerobot. Any idea(s) on where we should look to fix this?

vmayoral commented:
A few:
- implement a filter, that did it mostly for me
- don’t use default configs of the robot, instead leverage restricted degrees/s configs
- ensure proper setup conditions

curious about your use case. Feel free to reach out and if you can elaborate more on your use, happy to find some time and help.

srisaripalli commented:
> A few:
> 
>     * implement a filter, that did it mostly for me
> 
>     * don’t use default configs of the robot, instead leverage restricted degrees/s configs
> 
>     * ensure proper setup conditions
> 
> 
> curious about your use case. Feel free to reach out and if you can elaborate more on your use, happy to find some time and help.

Currently just trying to figure out if we can use an xARM 850 in an imitation learning framework (btw cool office:-)). If we get it working, we want to use both xARM's to move things around in a home like environment (dual arm config). Commenting out the max_relative_target seems to make the teleop OK but then replay still has issues.... Short video of replaying an episode here (https://youtube.com/shorts/Fzs2_MnPVe0)

imstevenpmwork commented:
I think this would be a great opportunity to test this new feature: https://github.com/huggingface/lerobot/pull/2123

imstevenpmwork commented:
Thank you for this hackathon contribution and the detailed demos. The PR targets the old robot_devices/manipulator.py + yaml-config + control_robot.py stack, all of which was removed in the Hardware API redesign (#777, e23b41e79) and config simplification (#550) — 12 of the 14 touched files no longer exist on main, and the HRI hints were patched into that deleted control loop (main now has voice announcements via log_say in lerobot-record). The way to bring UFACTORY 850 support today is a lerobot_robot_ufactory plugin package (docs/source/integrate_hardware.mdx, enabled by #2123) or a Robot subclass PR against the new API. Closing as obsolete — a rewrite as a plugin would be very welcome.
